hibike/hibike_communicator.py

```diff
@@ -0,0 +1,218 @@
+         If it's a device disconnection from a device whose uid in self.uids, the uid will be removed from self.uids
+         If it's a device value, cache it in the dictionary
+      """
+      while True:
+         output = self.stateQueue.get()
+
+         #Now, get or remove the uid if it is appropriate to do so
+         command = output[0]
+         if command == "device_subscribed":
+            uid = output[1][0]
+            if uid not in self.uids:
+               self.uids.add(uid)
+         if command == "device_disconnected":
+            uid = output[1]
+            if uid in self.uids:
+               self.uids.remove(uid)
+
+   
+         #If it is a device value, cache it in the dictionary.
+         if command == "device_values":
+             uid = output[1][0]
+             params_values = output[1][1]
+             for param_val_tuple in params_values:
+                if uid not in self.device_values_cache:
+                   self.device_values_cache.update({uid: {param_val_tuple[0]: (param_val_tuple[1], time.time())}})  
+                else:
+                   self.device_values_cache[uid].update({param_val_tuple[0]: (param_val_tuple[1], time.time())})
+   
+   def get_last_cached(self, uid, param):
+      """
+         Returns a tuple of the value and the timestamp of the last device_values package recieved from a uid and a parameter
+         Precondition: a device_data must have been recieved from the param before calling this function
+      """
+      return self.device_values_cache[uid][param]
+
+   
+   def get_uids_and_types(self):
+      """
+         Returns a list of tuples of all of the uids of all devices that the HibikeCommunicator talks to
+         Tuple structure: (uid, device type name)
+      """
+      list = []
+      for uid in self.uids:
+         list.append((uid, hm.uid_to_device_name(uid)))
+      return list
+
+   def write(self, uid, params_and_values):
+       """
+          Sends a Device Write to a device
+
+          uid - the device's uid
+          params_and_values - an iterable of param (name, value) tuples
+       """
+       self.pipeToChild.send(["write_params", [uid, params_and_values]])
+   
+   def read(self, uid, params): 
+        """
+           Sends a Device Read to a device
+        
+           uid - the device's uid
+           params - an iterable of the names of the params to be read
+        """
+        self.pipeToChild.send(["read_params", [uid, params]])
+        
+   def subscribe(self, uid, delay, params):
+        """
+           Subscribes to the device
+           
+           uid - the device's uid
+           delay - the delay between device datas to be sent
+           params - an iterable of the names of the params to be subscribed to
+        """
+        self.pipeToChild.send(["subscribe_device", [uid, delay, params]])
+
+##############
+#  TESTING   #
+##############
+def device_comms(comms, uid_type_tuple):
```

Comment:
Let's separate the two actions in each if block: Sending commands and printing the most current data of all live sensors each at regular (different) intervals. I recommend just copying from hibike_process's main function because it's nice and clean which == easier to modify.

Reply:
I will push my most recent edits of hibike/testing soon, might be useful as reference when switching to hibike_process's methods.